### PR TITLE
Add test report for Karmada1.3

### PR DESCRIPTION
**Pull request by Poor12**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
Add test report for Karmada1.3

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
None

### CODE REVIEW

File: blog/2022-10-26-test-report/index.md [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+3. Cluster Size: Cluster size is a dimension that cannot be ignored when measuring the resource pool size of a multi-cluster system. On the one hand, when the number of clusters is equal, the larger the scale of a single cluster, the larger the resource pool of the entire multi-cluster system.
+   On the other hand, the upper-layer feature of a multi-cluster system depends on the system's resource portrait of the cluster. For example, in the scheduling process, the resource status of the destination cluster is an indispensable factor.
+   Under the same circumstances, the larger the cluster scale, the greater the pressure on the control plane. Among cluster resources, Node and Pod are undoubtedly the two most important resources.
+   Node is the smallest carrier of computing, storage and other resources, and the number of Pods represents the application carrying capacity of a cluster. 
+   In fact, API objects in a single cluster also include common objects such as service, configmap, secret and so on. The introduction of these variables will make the testing process more complicated, so this test does not focus too much on the above variables.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:49]:
```suggestion
   In fact, API objects in a single cluster also include common objects such as Services, ConfigMaps, secrets and so on. The introduction of these variables will make the testing process more complicated, so this test does not focus too much on the above variables.
```

File: blog/2022-10-26-test-report/index.md [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+   * Num of Nodes
+   * Num of Pods
+
+For a multi-cluster system, it is impossible for a Kubernetes cluster to expand resource objects without limitation while satisfying SLIs/SLOs.
+Dimensions are sometimes not independent. As you move farther along one dimension, your cross-section wrt other dimensions gets smaller.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:54]:
```suggestion
Dimensions are interactive. One weighs more, the rest weigh less.
```

File: blog/2022-10-26-test-report/index.md [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+Taking the two dimensions of #Num of Clusters and #Num of Nodes as an example, in the scenario of stretching 5k nodes in a single cluster to 10k nodes under 100 clusters, or expanding the number of clusters to 200 clusters while maintaining the same size of a single cluster, the specifications of other dimensions are bound to be affected.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:55]:
```suggestion
Take the dimensions #Num of Clusters and #Num of Nodes as an example. When scaling from 5k nodes to 10k in 100 individual clusters, or expanding to 200 clusters while maintaining the original size of each cluster, the specifications of other dimensions will be surely affected.
```

File: blog/2022-10-26-test-report/index.md [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+| Offical | P99 latency of non-streaming read-only API calls (including GET and LIST) in the last 5 minutes                          | (a)Scope=resource, P99 <= 1s, (b)Scope=namespace or Scope=cluster, P99 <= 30s |
+
+2. Resource Distribution Latency
+
+
+| Status  | SLI                                                                                                                                                                                                                                                                         | SLO       |
+|---------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|
+| Offical | After the user submits the resource template and delivers the policy on the federated control plane, the P99 delay from the time when the resource is created on the member cluster, regardless of the network fluctuation between the control plane and the member cluster | P99 <= 2s |
+
+3. Cluster Register Latency
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:78]:
```suggestion
3. Cluster Registration Latency
```

File: blog/2022-10-26-test-report/index.md [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+
+
+| Status | SLI                                                                                                                                    | SLO |
+|--------|----------------------------------------------------------------------------------------------------------------------------------------|-----|
+| WIP    | P99 delay from when the cluster is registered in the federation control plane to when the status can be collected by the control plane | TBD |
+
+4. Resource usage
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:85]:
```suggestion
4. Resource Usage
```

File: blog/2022-10-26-test-report/index.md [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+
+
+| Status | SLI                                                                                                                                               | SLO |
+|--------|---------------------------------------------------------------------------------------------------------------------------------------------------|-----|
+| WIP    | The amount of necessary resource usage for the cluster federation to maintain its normal operation after registering a certain number of clusters | TBD |
+
+**Note**:
+
+1. The metrics measured do not account for network fluctuations between the control plane and member clusters. Also, Intra-cluster SLOs are not taken into account.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:94]:
```suggestion
1. Network fluctuations between the control plane and member clusters, as well as intra-cluster SLOs, are not taken into account for the metrics measured.
```

File: blog/2022-10-26-test-report/index.md [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+2. Resource Usage is a very important indicator of a multi-cluster system, but the upper-layer services provided by each system are different, so the resource requirements are also different. We do not make hard constraints here.
+3. Cluster Register Latency contains the duration from when the cluster is registered in the control plane to when the cluster becomes available.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:96]:
```suggestion
3. Cluster Registration Latency contains the duration from when the cluster is registered in the control plane to when the cluster becomes available.
```

File: blog/2022-10-26-test-report/index.md [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+
+### ClusterLoader2
+
+ClusterLoader2 is an open source Kubernetes cluster performance test tool. It can test the Kubernetes SLIs/SLOs to check whether the cluster meets the service quality standards. It also visualizes data for locating cluster problems and optimizing cluster performance.
+After the test, users get a performance report with detailed test results. However, ClusterLoader2 is a test tool customized for a single Kubernetes cluster, and in multi-cluster scenarios, it cannot obtain the resources of all clusters.
+Here we only use ClusterLoader2 to distribute resources.
+
+### Prometheus
+
+Prometheus is a systems and service monitoring system. It collects metrics from configured targets at given intervals, evaluates rule expressions, displays the results, and can trigger alerts when specified conditions are observed. After analyzing the processing of indicators by ClusterLoader2, we use Prometheus to monitor various indicators of the control plane according to specific query statements.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:109]:
```suggestion
Prometheus is a system and service monitoring system. It collects metrics from configured targets at given intervals, evaluates rule expressions, displays the results, and triggers alerts when specified conditions are observed. After analyzing the processing of indicators by ClusterLoader2, we use Prometheus to monitor various indicators of the control plane according to specific query statements.
```

File: blog/2022-10-26-test-report/index.md [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+
+![readonly latency namespace](./img/readonly_namespace.png)
+
+![mutating latency](./img/mutating.png)
+
+![readonly latency aggregated apiserver](./img/readonly_aggregated_apiserver.png)
+
+![mutating latency aggregated apiserver](./img/mutating_aggregated_apiserver.png)
+
+Cluster register latency:
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:360]:
```suggestion
Cluster registration latency:
```

File: blog/2022-10-26-test-report/index.md [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+
+
+| metric           | P50(ms) | P90(ms) | P99(ms) | SLO |
+|------------------|---------|---------|---------|-----|
+| cluster_register | 5356    | 6125    | 6904    | N/A |
+
+**Note**: Karmada's Pull mode is suitable for private cloud scenarios. Compared with Push mode, there is a component named karmada-agent running on the member cluster side. It will pull the user-submitted application and run it locally.
+In the process of Pull mode cluster registration, it will include the installation time of Karmada-agent in the member cluster. When the image of karmada-agent is ready, it will depend on the latency of Pod startup within a single cluster.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:368]:
```suggestion
In the process of cluster registration in Pull mode, the latency will include the installation time of Karmada-agent in the member cluster. When the image of karmada-agent is ready, the latency depends on that of Pod startup in a single cluster.
```

File: blog/2022-10-26-test-report/index.md [hunk 11]
```diff
@@ -0,0 +1,468 @@
...
+
+Resource Usage:
+
+![resource total](./img/pull_total.png)
+
+![cpu](./img/pull_cpu.png)
+
+![memory](./img/pull_memory.png)
+
+Karmada-agent in the member cluster with 5k Node and 2w Pod costs 40m CPU(cores) and 266Mi Memory(bytes).
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:408]:
```suggestion
Karmada-agent in the member cluster with 5k Nodes and 2w Pods consumes 40m CPU(cores) and 266Mi Memory(bytes).
```

File: blog/2022-10-26-test-report/index.md [hunk 12]
```diff
@@ -0,0 +1,468 @@
...
+
+## Conclusion and Analysis
+
+From the preceding test results, the API Call Latency and Resource Distribute Latency meet the SLIs/SLOs above. And the resources consumed by the system during the whole process are in a controllable range.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:412]:
```suggestion
From the preceding test results, the API Call Latency and Resource Distribution Latency meet the SLIs/SLOs above, and the resources consumed by the system during the whole process are in a controllable range.
```

File: blog/2022-10-26-test-report/index.md [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+Therefore, **Karmada can stably support 100 clusters with 500000 nodes and more than two million pods**. In production, Karmada effectively supports the management of hundreds of huge clusters.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:413]:
```suggestion
Therefore, **Karmada can stably support 100 clusters with 500,000 nodes and more than two million pods**. In production, Karmada effectively supports the management of hundreds of large-sized clusters.
```

File: blog/2022-10-26-test-report/index.md [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+Next, we will analyze the data of each indicator in detail.
+
+### Separation of Concerns: Resource Templates and Policies
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:416]:
```suggestion
### Separation of Concerns: Resource Templates and Policies
```
```suggestion
### Resource Templates and Policies
```

File: blog/2022-10-26-test-report/index.md [hunk 15]
```diff
@@ -0,0 +1,468 @@
...
+
+Karmada uses Kubernetes Native APIs for federated resource templates and reusable Policy APIs for cluster scheduling strategies. It not only makes it integrate with the Kubernetes ecosystem, but also greatly reduces the number of resources on the control plane.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:418]:
```suggestion
Karmada uses Kubernetes native APIs for federated resource templates and reusable Policy APIs for cluster scheduling strategies. These allow Karmada to integrate with the Kubernetes ecosystem and greatly reduce the number of resources on the control plane.
```

File: blog/2022-10-26-test-report/index.md [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+Based on this, the number of resources for the control plane does not depend on how many clusters on the control plane, but depends on numbers of multi-cluster applications.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:419]:
```suggestion
Based on this, the number of resources for the control plane does not depend on how many clusters on the control plane, but on how many multi-cluster applications.
```

File: blog/2022-10-26-test-report/index.md [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+
+Karmada's arthrecture inherits the simplicity and scalability of Kubernetes. Karmada-apiserver as the entry point of the control plane is like kube-apiserver in Kubernetes. You can optimize the component in multi-cluster scenarios using the parameters you need in single cluster configuration.
+
+During the entire resource distribution process, API call latency is within a resonable range.
+
+### Cluster Registry and Resource Distribution
+
+In Karmada 1.3, we provides the ability to register Pull mode clusters based on Bootstrap tokens, which not only simplifies the cluster registration process, but also facilitates access control.
+Now whether it is Push mode or Pull mode, we both can use the karmadactl CLI to complete the cluster registration. Compared with Push mode, there is a component named karmada-agent running on the member cluster side.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:428]:
```suggestion
For both the Push and Pull modes, we can use the karmadactl CLI to complete cluster registration. Unlike the Push mode, the Pull mode runs a component named karmada-agent on member clusters.
```

File: blog/2022-10-26-test-report/index.md [hunk 18]
```diff
@@ -0,0 +1,468 @@
...
+
+Cluster registry latency above contains the delay for the control plane to successfully collect the member cluster status. In the process of cluster lifecycle management, Karmada will collect the cluster version, API enablement and whether the cluster is online or healthy.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:430]:
```suggestion
Cluster registration latency contains the latency for the control plane to successfully collect the member cluster status. In the process of cluster lifecycle management, Karmada will collect the cluster version, API enablement and whether the cluster is online or healthy.
```

Comment (Tingtal) [reply blog/2022-10-26-test-report/index.md:430]: Cluster **Registry** Latency or Cluster **Registration** Latency, please choose one and use it in all the places it shows up.

File: blog/2022-10-26-test-report/index.md [hunk 19]
```diff
@@ -0,0 +1,468 @@
...
+Besides, Karmada will obtain the resource usage of the cluster to model the cluster so that the scheduler can better select the target cluster. In this case, Cluster Registry Latency depends on the size of the cluster.
+This article shows the latency of joining a 5,000-node cluster until it is ready. You can [disable cluster resource modeling](https://karmada.io/docs/next/userguide/scheduling/cluster-resources#disable-cluster-resource-modeling) so that cluster registry latency will be independent of cluster size and less than 2000ms.
+
+Whether Push mode or Pull mode, Karmada maintains a fast rate for distributing resources to member clusters. The only difference is that karmada-controller-manager on the control plane is responsible for the distribution of all cluster resources in Push mode whereas karmada-agent is only responsible for one cluster in Pull mode.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:434]:
```suggestion
No matter whether you are using the Push mode or Pull mode, Karmada maintains a fast rate for distributing resources to member clusters. The only difference is that karmada-controller-manager on the control plane is responsible for the distribution of all cluster resources in Push mode whereas karmada-agent is only responsible for one cluster in Pull mode.
```

File: blog/2022-10-26-test-report/index.md [hunk 20]
```diff
@@ -0,0 +1,468 @@
...
+Therefore, in the process of distributing resources with high concurrency, Pull mode will be slightly faster than Push mode under the same configuration.
+Also, you may adjust `--concurrent-work-syncs` parameter in karmada-controller-manager to improve performance.
+
+### Resource Usage between Push mode and Pull mode
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:438]:
```suggestion
### Resource Usage between Push Mode and Pull Mode
```

File: blog/2022-10-26-test-report/index.md [hunk 21]
```diff
@@ -0,0 +1,468 @@
...
+
+In Karmada 1.3, we do a lot of work to reduce Karmada's resource consumption when managing huge clusters. Now we are happy to announce that compared with version 1.2, Karmada 1.3 reduces memory consumption by **85%** and CPU consumption by **32%** in large-scale scenarios.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:440]:
```suggestion
In Karmada 1.3, we did a lot to reduce Karmada's resource consumption when managing large clusters. Compared with version 1.2, Karmada 1.3 reduces memory consumption by **85%** and CPU consumption by **32%** in large-scale scenarios.
```

File: blog/2022-10-26-test-report/index.md [hunk 22]
```diff
@@ -0,0 +1,468 @@
...
+In general, Pull Mode has a clear advantage in memory usage, and the rest of the resources are similar.
+
+In Push mode, the main resource consumption of the control plane is concentrated on the karmada-controller-manager. And karmada-apiserver is not under much pressure.
+
+![push apiserver qps](./img/push_apiserver_qps.png)
+
+From the qps of karmada-apiserver and the request latency of karmada-etcd, we can know that the number of requests to karmada-apiserver is kept at a low level.
+In Push mode, all requests come from karmada-controller-manager on the control plane.  You can configure `--kube-api-qps` and `--kube-api-burst` in karmada-controller-manager to control requests within a certain threshold.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:448]:
```suggestion
In Push mode, all requests come from karmada-controller-manager on the control plane. You can configure `--kube-api-qps` and `--kube-api-burst` in karmada-controller-manager to control requests within a certain threshold.
```

File: blog/2022-10-26-test-report/index.md [hunk 23]
```diff
@@ -0,0 +1,468 @@
...
+In Pull mode, the main resource consumption of the control plane is concentrated on the karmada-apiserver instead of karmada-controller-manager.
+
+![pull apiserver qps](./img/pull_apiserver_qps.png)
+
+From the qps of karmada-apiserver and the request latency of karmada-etcd, we can know that the number of requests to karmada-apiserver is kept at a high level.
+In Pull mode, karmada-agent of each member cluster needs to maintain a long connection with karmada-apiserver.
+We can easily conclude that the number of requests for karmada-apiserver will be N times that of configuration in karmada-agent(N=#Num of clusters).
+Therefore, in the scenario of a large number of Pull mode clusters, we recommand increasing the `--max-requests-inflight`, `--max-mutating-requests-inflight` in karmada-apiserver and `--quota-backend-bytes` in karmada-etcd to improve the throughout of the control plane.
+
+Now Karmada provides the ability named [cluster resource modeling](https://karmada.io/docs/next/userguide/scheduling/cluster-resources) to make scheduling decisions in scenario of dynamic replica assignment based on cluster free resources.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:459]:
```suggestion
Now Karmada provides the ability named [cluster resource modeling](https://karmada.io/docs/next/userguide/scheduling/cluster-resources) to make scheduling decisions in scenario of dynamic replica assignment based on idle cluster resources.
```

File: blog/2022-10-26-test-report/index.md [hunk 24]
```diff
@@ -0,0 +1,468 @@
...
+In the process of resource modeling, it will collect node and pod information from all clusters managed by Karmada. This imposes a considerable performance burden in large-scale scenarios.
+If you do not use this ability, you can [disable cluster resource modeling](https://karmada.io/docs/next/userguide/scheduling/cluster-resources#disable-cluster-resource-modeling) to further reduce resource consumption.
+
+### Summary
+
+In terms of use scenarios, Push mode is suitable for managing Kubernetes clusters on public clouds, while Pull mode covers private cloud and edge-related scenarios compared to Push mode. In terms of performance and security, the overall performance of the Pull mode is better than that of the Push mode.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:465]:
```suggestion
In terms of use scenarios, Push mode is suitable for managing Kubernetes clusters on public clouds, while Pull mode covers private cloud and edge-related scenarios. In terms of performance and security, the Pull mode outperforms the Push mode. 
```

File: blog/2022-10-26-test-report/index.md [hunk 25]
```diff
@@ -0,0 +1,468 @@
...
+In the process of resource modeling, it will collect node and pod information from all clusters managed by Karmada. This imposes a considerable performance burden in large-scale scenarios.
+If you do not use this ability, you can [disable cluster resource modeling](https://karmada.io/docs/next/userguide/scheduling/cluster-resources#disable-cluster-resource-modeling) to further reduce resource consumption.
+
+### Summary
+
+In terms of use scenarios, Push mode is suitable for managing Kubernetes clusters on public clouds, while Pull mode covers private cloud and edge-related scenarios compared to Push mode. In terms of performance and security, the overall performance of the Pull mode is better than that of the Push mode. 
+Each cluster is managed by the karmada-agent component in the cluster and is completely isolated.
+However, while the Pull mode improves performance, it also needs to correspondingly improve the performance of karmada-apiserver and karmada-etcd to face challenges in high-traffic and high-concurrency scenarios. For specific methods, please refer to the optimization of kubernetes for large-scale clusters.
```
Comment (Tingtal) [inline blog/2022-10-26-test-report/index.md:467]:
```suggestion
However, while the Pull mode improves performance, it also needs to correspondingly improve the performance of karmada-apiserver and karmada-etcd to face challenges in traffic-intensive and high-concurrency scenarios. For specific methods, please refer to the optimization of Kubernetes for large-scale clusters.
```

**Review by Tingtal — CHANGES_REQUESTED** [review]
Language check.

**Comment by RainbowMango** [comment]
Please check and close the comments made by reviewers. And leave your comments on items that you don't agree with.

File: blog/2022-10-26-test-report/index.md [hunk 26]
```diff
@@ -0,0 +1,468 @@
...
+
+The Karmada Control Plane consists of the following components:
+
+- Karmada API Server
+- Karmada Controller Manager
+- Karmada Scheduler
+
+ETCD stores the Karmada API objects, the API Server is the REST endpoint all other components talk to, and the Karmada Controller Manager performs operations based on the API objects you create through the API server.
+
+The Karmada Controller Manager runs the various controllers,  the controllers watch Karmada objects and then talk to the underlying clusters' API servers to create regular Kubernetes resources.
```
[the remark this answers is not in the document]
Comment (Poor12) [reply blog/2022-10-26-test-report/index.md:32]: Pasted from README.md in karmada.

**Comment by Poor12** [comment]
> Please check and close the comments made by reviewers. And leave your comments on items that you don't agree with.

Done

File: blog/2022-10-26-test-report/index.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply blog/2022-10-26-test-report/index.md:32]: no big deal I guess.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**